### PR TITLE
API Documentation

### PR DESCRIPTION
**Pull request by jsancio**
- [x] Remove the generate `api.html`
- [x] File two Jiras for both TODOs.
- [x] Remove TODOs from the PR.

### CODE REVIEW

**Comment by jsancio** [comment]
@cruhland and @larioj I'll leave the HTML file for now to help during the review. I will remove it before merging the code.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 1]
```diff
@@ -0,0 +1,323 @@
+#%RAML 1.0
+title: DC/OS Package and Service
+/package:
+  description: Implements all packaging operation required for managing packages in DC/OS.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: operation -> operations

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+  /install:
+    description: Runs a service from a package from Universe.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: package from Universe -> Universe package

Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Should the description be under `/install` or under the `post` entry below? We might decide to use another verb in the future?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+    post:
+      headers:
+        Accept:
+          enum: [application/vnd.dcos.package.install-response+json;charset=utf-8;version=v2]
+      body:
+        application/vnd.dcos.package.install-request+json;charset=utf-8;version=v1:
+          type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/InstallRequest-schema.json
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: This huge relative path is a bit annoying, but I'm guessing there's no other way to do it?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+        Accept:
+          enum: [application/vnd.dcos.package.uninstall-response+json;charset=utf-8;version=v1]
+      body:
+        application/vnd.dcos.package.uninstall-request+json;charset=utf-8;version=v1:
+          type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/UninstallRequest-schema.json
+          example:
+            packageName: cassandra
+            appId: /cassandra
+      responses:
+        200:
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Only documenting the success responses for now?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+          body:
+            application/vnd.dcos.package.uninstall-response+json;charset=utf-8;version=v1:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/UninstallResponse-schema.json
+              example:
+                results:
+                  - packageName: cassandra
+                    appId: /cassandra
+  /describe:
+    description: |
+      Show information about the given package including the resource need to start the service,
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: need -> needed

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+      configure to start the service and command line extensions included with the package.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: configure -> configuration/config?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        200:
+          body:
+            application/vnd.dcos.package.describe-response+json;charset=utf-8;version=v2:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v2/model/DescribeResponse-schema.json
+              example:
+                name: cassandra
+                version: 1.0.24-3.0.10
+                packagingVersion: "3.0"
+                releaseVersion: 24
+                maintainer: support@mesospher.io
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: mesospher.io -> mesosphere.io

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+                description: Apache Cassandra running on DC/OS
+                tags:
+                  - data
+                  - database
+                  - nosql
+  /render:
+    description: |
+      Generates the Marathon Application Definition for the given package name and version, and an
+      optional configuration JSON object.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: The phrasing made me think for a moment that the endpoint generated the app definition *and* the configuration JSON. Maybe use the following?

> Generates the Marathon Application Definition for the given package name, version, and optional configuration JSON object.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 9]
```diff
@@ -0,0 +1,323 @@
...
+            application/vnd.dcos.package.list-versions-response+json;charset=utf-8;version=v1:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/ListVersionsResponse-schema.json
+              example:
+                results:
+                  1.0.24-3.0.10: "24"
+                  1.0.23-3.0.10: "23"
+  /search:
+    description: |
+      Lists all matching packages in the repository given a partial pattern. The character \'*\'
+      can be used to match any number of characters.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Should this mention that we search the package name, description, and tags fields?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 10]
```diff
@@ -0,0 +1,323 @@
...
+                    versions:
+                      1.0.24-3.0.10: "24"
+                      1.0.23-3.0.10: "23"
+                    description: Apache Cassandra
+                    framework: true
+                    tags:
+                      - database
+  /list:
+    description: |
+      Lists all of the running of the running DC/OS Services started from a DC/OS Package.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Remove duplicate "of the running".

Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Did you want to mention that you can filter by package name and/or app ID?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 11]
```diff
@@ -0,0 +1,323 @@
...
+          enum: [application/vnd.dcos.package.add-response+json;charset=utf-8;version=v1]
+      body:
+        application/vnd.dcos.package.add-request+json;charset=utf-8;version=v1:
+          type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/PackageAddRequest-schema.json
+          example:
+            packageName: cassandra
+            packageVersion: 1.0.24-3.0.10
+        application/vnd.dcos.universe.package+zip;version=v1:
+      responses:
+        200: # TODO: confirm this is 200
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Don't forget this!

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 12]
```diff
@@ -0,0 +1,323 @@
...
+          body:
+            application/vnd.dcos.package.add-response+json;charset=utf-8;version=v1:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/PackageAddResponse-schema.json
+              example:
+                name: cassandra
+                version: 1.0.24-3.0.10
+                packagingVersion: "3.0"
+                releaseVersion: 24
+                maintainer: support@mesospher.io
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: mesospher.io -> mesosphere.io

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 13]
```diff
@@ -0,0 +1,323 @@
...
+                description: Apache Cassandra running on DC/OS
+                tags:
+                  - data
+                  - database
+                  - nosql
+  /repository:
+    /list:
+      description: |
+        List all of the package repository (i.e Universe) used by DC/OS. The order in the list is
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: repository -> repositories

Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Would "e.g." ("for example") make more sense than "i.e." ("in other words")?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 14]
```diff
@@ -0,0 +1,323 @@
...
+        the order in which packages are found.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: How about this for the second sentence?

> The list is ordered by priority; if multiple repositories contain the same package, the one closest to the start of the list will be used.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 15]
```diff
@@ -0,0 +1,323 @@
...
+            body:
+              application/vnd.dcos.package.repository.list-response+json;charset=utf-8;version=v1:
+                type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/PackageRepositoryListResponse-schema.json
+                example:
+                  repositories:
+                    - name: Universe
+                      uri: https://universe.mesosphere.com/repo
+    /add:
+      description: |
+        Adds a package repository (i.e. Universe) for use by DC/OS. To add a package repository to
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: i.e. vs. e.g.?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 16]
```diff
@@ -0,0 +1,323 @@
...
+        the beginning of list set the index to zero (0). To add a package repository to the end of
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: "of list" -> "of the list"

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 17]
```diff
@@ -0,0 +1,323 @@
...
+                type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/PackageRepositoryAddResponse-schema.json
+                example:
+                  repositories:
+                    - name: Universe
+                      uri: https://universe.mesosphere.com/repo
+                    - name: Local Universe
+                      uri: http://universe.marathon.mesos:8080/repo
+    /delete:
+      description: |
+        Deletes a package repository (i.e. Universe) from DC/OS.
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: i.e. vs. e.g.?

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 18]
```diff
@@ -0,0 +1,323 @@
...
+        Accept:
+          enum: [application/vnd.dcos.sevice.start-response+json;charset=utf-8;version=v1]
+      body:
+        application/vnd.dcos.sevice.start-request+json;charset=utf-8;version=v1:
+          type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/ServiceStartRequest-schema.json
+          example:
+            packageName: cassandra
+            packageVersion: 1.0.24-3.0.10
+      responses:
+        200: # TODO: confirm this is 200
```
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Don't forget this!

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 2 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Sure. I'll move it into the method section.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 3 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Not that I know of. The basic problem is that the schema are part of the `cosmos-model` while the api route description are part of the `cosmos-server` module.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 4 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Yep. Error cases are going to be difficult to document.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 8 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Done.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 9 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Maybe. Let's leave it like this for now. I don't want to get into the implementation details for such a action like "search". For example search engine don't tell you how they work. We can fix it if users need more details.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 10 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: It is easier to document what the properties do in the request body schema.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 13 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Done and done.

**Comment by jsancio** [comment]
@cruhland I moved the description to be associated with the method. I also included all of your type and grammatical corrections.

I am going to work on removing the TODOs regarding the return status for some of the of the responses.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 9 above)
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: :+1:

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 10 above)
Comment (cruhland) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: :+1:

**Comment by jsancio** [comment]
@cruhland Fixed the return status for `package/add` and removed the TODOs.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+          body:
+            application/vnd.dcos.package.install-response+json;charset=utf-8;version=v2:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v2/model/InstallResponse-schema.json
+              example:
+                packageName: cassandra
+                packageVersion: 1.0.24-3.0.10
+                appId: /cassandra
+  /uninstall:
+    post:
+      description: Stops a running service that was launched from a DC/OS Package.
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: s/s `DC/OS package`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 20]
```diff
@@ -0,0 +1,321 @@
...
+            application/vnd.dcos.package.uninstall-response+json;charset=utf-8;version=v1:
+              type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/UninstallResponse-schema.json
+              example:
+                results:
+                  - packageName: cassandra
+                    appId: /cassandra
+  /describe:
+    post:
+      description: |
+        Show information about the given package including the resources need to start the service,
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `Show information about the package, including the required resources and configuration to start the service, and command line extensions that are included with the package.`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 21]
```diff
@@ -0,0 +1,321 @@
...
+                maintainer: support@mesosphere.io
+                description: Apache Cassandra running on DC/OS
+                tags:
+                  - data
+                  - database
+                  - nosql
+  /render:
+    post:
+      description: |
+        Generates the Marathon Application Definition for the given package name, version, and
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `Generates the Marathon application definition for the package name, version, and optional configuration JSON object.`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 22]
```diff
@@ -0,0 +1,321 @@
...
+                      1.0.24-3.0.10: "24"
+                      1.0.23-3.0.10: "23"
+                    description: Apache Cassandra
+                    framework: true
+                    tags:
+                      - database
+  /list:
+    post:
+      description: |
+        Lists all of the running DC/OS Services started from a DC/OS Package.
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `Lists all of the running DC/OS services started from a DC/OS package.`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 23]
```diff
@@ -0,0 +1,321 @@
...
+                        version: 1.0.24-3.0.10
+                        maintainer: support@mesosphere.io
+                        description: Apache Cassandra
+                        tags:
+                          - database
+
+  /add:
+    post:
+      description: |
+        **Experimental.** Adds a DC/OS Package to DC/OS. To add a DC/OS Package from Universe use
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `Adds a DC/OS package to DC/OS. To add a DC/OS package ...`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 24]
```diff
@@ -0,0 +1,321 @@
...
+                tags:
+                  - data
+                  - database
+                  - nosql
+  /repository:
+    /list:
+      post:
+        description: |
+          List all of the package repositories (for example Universe) used by DC/OS. The list is
+          ordered by priority; if multiple repositories contain the same package, the one closest
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `...the same package, the package closest...`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 25]
```diff
@@ -0,0 +1,321 @@
...
+            body:
+              application/vnd.dcos.package.repository.delete-response+json;charset=utf-8;version=v1:
+                type: !include ../../../../../../../../cosmos-model/src/main/resources/com/mesosphere/cosmos/rpc/v1/model/PackageRepositoryDeleteResponse-schema.json
+                example:
+                  repositories:
+                    - name: Universe
+                      uri: https://universe.mesosphere.com/repo
+/capabilities:
+  get:
+    description: List all of the capabilities supported by this DC/OS.
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: ` List all of the capabilities supported by this version DC/OS.` Not sure what `this` refers to.

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml [hunk 26]
```diff
@@ -0,0 +1,321 @@
...
+              capabilities:
+                - name: PACKAGE_MANAGEMENT
+                - name: SUPPORT_CLUSTER_REPORT
+                - name: METRONOME
+                - name: LOGGING
+/service:
+  /start:
+    post:
+      description: |
+        **Experimental.** Starts a DC/OS Service from a DC/OS Package.
```
Comment (joel-hamill) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Suggest: `Starts a DC/OS service from a DC/OS package.`

File: cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml (hunk 25 above)
Comment (jsancio) [inline cosmos-server/src/main/resources/com/mesosphere/cosmos/handler/api.raml]: Yeah. I change it to `... supported by DC/OS.`

**Comment by jsancio** [comment]
@joel-hamill Done.

**Pull request merged**